server/src/main/java/io/druid/client/SingleServerInventoryView.java

```diff
@@ -77,4 +94,58 @@ protected DruidServer removeInnerInventory(DruidServer container, String invento
     removeSingleInventory(container, inventoryKey);
     return container;
   }
+
+  @Override
+  public void registerSegmentCallback(
+      final Executor exec, final SegmentCallback callback, final Predicate<Pair<DruidServerMetadata, DataSegment>> filter
+  )
+  {
+    segmentPredicates.put(callback, filter);
+    registerSegmentCallback(
+        exec, new SegmentCallback()
+        {
+          @Override
+          public CallbackAction segmentAdded(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            final CallbackAction action;
+            if(filter.apply(Pair.of(server, segment))) {
+              action = callback.segmentAdded(server, segment);
+              if (action.equals(CallbackAction.UNREGISTER)) {
+                segmentPredicates.remove(callback);
+              }
+            } else {
+              action = CallbackAction.CONTINUE;
+            }
+            return action;
+          }
+
+          @Override
+          public CallbackAction segmentRemoved(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            {
+              final CallbackAction action;
+              if(filter.apply(Pair.of(server, segment))) {
+                action = callback.segmentRemoved(server, segment);
+                if (action.equals(CallbackAction.UNREGISTER)) {
+                  segmentPredicates.remove(callback);
+                }
+              } else {
+                action = CallbackAction.CONTINUE;
+              }
+              return action;
+            }
+          }
+
+          @Override
+          public CallbackAction segmentViewInitialized()
+          {
+            return callback.segmentViewInitialized();
+          }
+        }
+    );
+  }
```

Comment:
this looks exactly same as the one in BatchServerInventoryView, can you refactor and reuse?